### DISCUSSION TITLE
lua gsub behavior inconsistent with luajit?

### DISCUSSION THREAD

**Discussion by xq114** [xmake-io/xmake#1724] [General]
https://stackoverflow.com/questions/45179381/lua-gsub-replace-with

`a:gsub("%%","")`在luajit里面合法，但在lua里面不合法而是要写`a:gsub("%%%%", "")`，写的时候以哪一种为准呢

**Comment by waruqi** [comment]
这个要回头细看下，尽量以luajit为准，做到向下兼容，毕竟lua 版本以后会一直升，有可能一些语法会一直变动，但是luajit的语法固定了 兼容性更好

回头我会内部适配下 lua 端做内部兼容，适配到luajit

另外尽量走 replace("%",{plain=true})，兼容性更好，也更快，跟是否lua/luajit无关

**Comment by waruqi** [comment]
> a:gsub("%%","")在luajit里面合法，但在lua里面不合法而是要写a:gsub("%%%%", "")，写的时候以哪一种为准呢

测了下，跟参数一没任何关系，lua和luajit的行为一致，都是下面的结果。

```console
xmake> ("foo%%bar"):gsub("%%%%", "_")
< 1: "foo_bar"
< 2: 1
xmake> ("foo%%bar"):gsub("%%", "_")
< 1: "foo__bar"
< 2: 2
```

其实是第二个参数带有 `%` 的转义行为不一致，lua 下 仅仅用 `%` 会报错，但是用 `%%` lua/luajit 的结果也一致

### lua

```console
xmake> ("foo%%bar"):gsub("%%", "%")
stdin:1: invalid use of '%' in replacement string
xmake> ("foo%%bar"):gsub("%%", "%%")
< 1: "foo%%bar"
< 2: 2
```

### luajit

```console
xmake> ("foo%%bar"):gsub("%%", "%")
< 1: "foobar"
< 2: 2
xmake> ("foo%%bar"):gsub("%%", "%%")
< 1: "foo%%bar"
< 2: 2
```

**Reply by waruqi** [reply]
```console
xmake> ("foo%%bar"):gsub("%%", "%")
< 1: "foobar"
< 2: 2
```

而 luajit 下，虽然允许第二个参数是 `%` ，但是替换的结果也是有问题的，

```console
xmake> a = ("foo%%bar"):gsub("%%", "%")
xmake> import("core.base.bytes").new(a)
< bytes(8) <0x66 0x6f 0x6f 0x00 0x00 0x62 0x61 0x72>
```

`foo``bar` 中间多了两个 `\0\0`，也没有正确替换成 `%`

也就是说，即使 luajit，第二个参数也应该设置 `%%` 做转义处理，lua/luajit 行为都是一致的。。只不过luajit下对 `%` 的参数2输入没做严格的检查和报错处理，直接就去替换了，也没保证正确性。

**Reply by waruqi** [reply]
所以，总结就是，gsub的参数一，参数二，如果需要处理 `%` 字符，不管是 lua/还是luajit 都是需要转义设置成 `%%`，行为一样。。只不过lua对非法的单个 `%` 做了更严格的检查，直接报错了，而不是强行做无效替换
